osquery/tables/system/posix/shell_history.cpp

```diff
@@ -12,76 +12,119 @@
 
 #include <osquery/core.h>
 #include <osquery/filesystem/filesystem.h>
-#include <osquery/logger.h>
 #include <osquery/tables.h>
-#include <osquery/tables/system/system_utils.h>
 #include <osquery/tables/system/posix/shell_history.h>
-#include <osquery/utils/conversions/split.h>
+#include <osquery/tables/system/system_utils.h>
 #include <osquery/utils/system/system.h>
 
 namespace osquery {
 namespace tables {
 
 const std::vector<std::string> kShellHistoryFiles = {
-    ".bash_history", ".zsh_history", ".zhistory", ".history", ".sh_history",
+    ".bash_history",
+    ".zsh_history",
+    ".zhistory",
+    ".history",
+    ".sh_history",
+};
+
+struct HistoryState {
+  std::string content;
+  std::regex bash_timestamp_rx{"^#([0-9]+)$"};
+  std::regex zsh_timestamp_rx{"^: {0,10}([0-9]{1,11}):[0-9]+;(.*)$"};
+  std::string prev_bash_timestamp;
 };
 
-void genShellHistoryFromFile(const std::string& uid,
-                             const boost::filesystem::path& history_file,
-                             QueryData& results) {
-  std::string history_content;
-  if (forensicReadFile(history_file, history_content).ok()) {
-    std::regex bash_timestamp_rx("^#([0-9]+)$");
-    std::regex zsh_timestamp_rx("^: {0,10}([0-9]{1,11}):[0-9]+;(.*)$");
-
-    std::string prev_bash_timestamp;
-    for (const auto& line : split(history_content, "\n")) {
-      std::smatch bash_timestamp_matches;
-      std::smatch zsh_timestamp_matches;
-
-      if (prev_bash_timestamp.empty() &&
-          std::regex_search(line, bash_timestamp_matches, bash_timestamp_rx)) {
-        prev_bash_timestamp = bash_timestamp_matches[1];
-        continue;
-      }
-
-      Row r;
-
-      if (!prev_bash_timestamp.empty()) {
-        r["time"] = INTEGER(prev_bash_timestamp);
-        r["command"] = line;
-        prev_bash_timestamp.clear();
-      } else if (std::regex_search(
-                     line, zsh_timestamp_matches, zsh_timestamp_rx)) {
-        std::string timestamp = zsh_timestamp_matches[1];
-        r["time"] = INTEGER(timestamp);
-        r["command"] = zsh_timestamp_matches[2];
-      } else {
-        r["time"] = INTEGER(0);
-        r["command"] = line;
-      }
-
-      r["uid"] = uid;
-      r["history_file"] = history_file.string();
-      results.push_back(r);
+void genShellHistoryFromFile(
+    const std::string& uid,
+    const boost::filesystem::path& history_file,
+    std::function<void(DynamicTableRowHolder& row)> predicate) {
+  struct HistoryState hState;
+
+  auto parseLine =
+      [&hState, &uid, &history_file, &predicate](std::string& line) {
+        std::smatch bash_timestamp_matches;
+        std::smatch zsh_timestamp_matches;
+
+        if (hState.prev_bash_timestamp.empty() &&
+            std::regex_search(
```

Comment:
Given that this is a performance oriented PR, I wonder if we should be keeping regex. My expectation is that it's a performance hit. It feels odd to run this even for places we know it doesn't apply.

Reply:
I am a huge fan of the "one thesis rule" and I realize I am violating that slightly with this PR by folding in several performance improvements. But in my defense the last and most critical change, implementing a generator-based yield, is the real thesis and the other changes are required dependency refactors.

That said, I like your suggestion and I recommend we do make changes to how the regexes are used. But I recommend we do so in a follow up PR.